### ISSUE TITLE
does it work on Fedora 22 ?

### ISSUE THREAD

**Issue by sirgogo** [FilipDominec/python-meep-install#1]
Hi,

I'm having lots of issues installing python-meep on Fedora 22. I succesfully installed MEEP --with-mpi option, but have had no luck installing python meep.

In particular, it fails when I do a ./make-mpi like this:

swig -python -c++ -w503,389,302,362,314,509,451 -I/usr/local/include -o meep_mpi_wrap.cpp meep_mpi.i
meep-site-init.py:1: Error: Unknown SWIG preprocessor directive: initialisations (if this is a block of target language code, delimit it with %{ and %})

Any ideas?

I was able to get it a bit further by changing Line1 of meep-site-init.py 
from: #initialisations specific for .... ....
to: /*#initialisations specific for .... ....*/

Output looks ilke this:

sudo ./make-mpi
Directories for SWIG -I option :  ['/usr/local/include']
Directories for -L option :  ['/usr/local/lib']
running clean
removing 'build/lib.linux-x86_64-2.7' (and everything under it)
'build/bdist.linux-x86_64' does not exist -- can't clean it
'build/scripts-2.7' does not exist -- can't clean it
removing 'build'
Directories for SWIG -I option :  ['/usr/local/include']
Directories for -L option :  ['/usr/local/lib']
running build_ext
building '_meep_mpi' extension
swigging meep_mpi.i to meep_mpi_wrap.cpp
swig -python -c++ -w503,389,302,362,314,509,451 -I/usr/local/include -o meep_mpi_wrap.cpp meep_mpi.i
creating build
creating build/temp.linux-x86_64-2.7
gcc -pthread -fno-strict-aliasing -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -DNDEBUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -fPIC -I/usr/local/include -I/usr/include/python2.7 -c meep_mpi_wrap.cpp -o build/temp.linux-x86_64-2.7/meep_mpi_wrap.o
In file included from /usr/include/numpy/ndarraytypes.h:1804:0,
                 from /usr/include/numpy/ndarrayobject.h:17,
                 from /usr/include/numpy/arrayobject.h:4,
                 from custom.hpp:19,
                 from meep_mpi_wrap.cpp:3937:
/usr/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it by " \
....

continues for some time, but finally says:
g++ -pthread -shared -Wl,-z,relro build/temp.linux-x86_64-2.7/meep_mpi_wrap.o -L/usr/local/lib -L/usr/lib64 -lmeep_mpi -lpython2.7 -o build/lib.linux-x86_64-2.7/_meep_mpi.so
/bin/ld: cannot find -lmeep_mpi
collect2: error: ld returned 1 exit status
error: command 'g++' failed with exit status 1

Any ideas? I have very little clue about how python-meep is put together, but I badly need this to work!!

Thanks in advance,

**Comment by FilipDominec** [comment]
Hi, first of all, thanks for using and testing the python-meep-install script! 

So far, I tested and fixed it for multiple versions of (L)Ubuntu 12.10-15.04 and for Debian Wheezy/Jessie (on both architectures i386 and x86_64). 

I did not test it on any RPM based distro so far, so I am going to do so today: I will try to repeat your procedure with the error and find out what goes wrong. Stay tuned.

**Comment by sirgogo** [comment]
Hello Filip,

Thanks for your help, very appreciated. I have included the set of commands
I used from a vanilla Fedora22 install. Hopefully you have better luck than
me!

Sincerely,
Abhe

**Comment by FilipDominec** [comment]
I installed Fedora 22 in a virtual machine and tried hard to adapt python-meep-install.sh to work also there. About a half of dependencies have a different package name, so I had to introduce DEB/RPM switches. After that, the meep's autogen command fails with "could not find mpi library for --with-mpi".

This is sad, as multiprocessing is useful and I tried to get OpenMPI working on Fedora. Anyway, I will try compiling without MPI at all.

**Comment by sirgogo** [comment]
I see. Well, single core is better than nothing!

Any clues as to where its failing? Perhaps.. meep wasn't compiled properly with MPI? I'm not sure how to check if it was made properly, but I didn't get any errors during the ./configure command.

**Comment by FilipDominec** [comment]
Commenting out the line https://github.com/FilipDominec/python-meep-install/blob/master/python-meep-install.sh#L6 and uncommenting line 9, I could run the whole compilation procedure in Fedora 22. So it works for me (except for failing at some HDF5 export during the simulatino, which is a slightly different issue).

If you run it on your system, you can save the log using ./python-meep-utils.sh | tee fedora22gogo.log
Can you then compare your output with mine? What are the differences?

**Comment by sirgogo** [comment]
Hmm.. its largely the same. I did notice some differences.  Still, it seems to go through most but ends in an error:

g++ -pthread -shared -Wl,-z,relro build/temp.linux-x86_64-2.7/meep_wrap.o -L/usr/local/lib -L/usr/lib64 -lmeep -lpython2.7 -o build/lib.linux-x86_64-2.7/_meep.so
/bin/ld: /usr/local/lib/libmeep.a(anisotropic_averaging.o): relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/usr/local/lib/libmeep.a: error adding symbols: Bad value
collect2: error: ld returned 1 exit status
error: command 'g++' failed with exit status 1

Any ideas?

**Comment by FilipDominec** [comment]
Your latest error is of a different sort than any of the previous; it seems to me that you just omitted the -fPIC option for the compiler to generate Position Independent Code on an x86_64 machine.

**Comment by sirgogo** [comment]
So, where would I append that option flag? It seems that python-meep-install grabs and installs its own meep, independent of the one I made before-- is this true?

**Comment by FilipDominec** [comment]
The option flag has to be passed as an option to the compiler; to be safe, I passed it to all compiler options and did not change it later. https://github.com/FilipDominec/python-meep-install/blob/master/python-meep-install.sh#L57 This should resolve the last issue. I think it would be very useful if you tried modifying the python-meep-install.sh file, so that we can easily merge your results.

To your question: yes, the script is set up to install meep and python-meep into the /usr/local/ directory so that it does not overwrite other versions.

**Comment by sirgogo** [comment]
Cool, I grabbed the updated script. However, on to the next error!

g++ -pthread -shared -Wl,-z,relro build/temp.linux-x86_64-2.7/meep_mpi_wrap.o -L/usr/local/lib -L/usr/lib64 -lmeep_mpi -lpython2.7 -o build/lib.linux-x86_64-2.7/_meep_mpi.so
/bin/ld: cannot find -lmeep_mpi
collect2: error: ld returned 1 exit status
error: command 'g++' failed with exit status 1

And, sure thing: I will add in options for Fedora just as soon as we can get this working!

**Comment by FilipDominec** [comment]
Though I always find these output a bit too cryptic, it appears the error originates from that the linker can not find the newly compiled meep_mpi.so library. 

Can you figure out where it is really installed? If you find some unusual path, it can be added to LDFLAGS after line 63.

**Comment by sirgogo** [comment]
Honestly, I could not figure it out on Fedora 22 yet. I will try later this summer. 

But until then, I needed an install going. So I came up with a work around: containerized meep with using docker. I've also installed python-meep, and it seems to work with openMPI (although there was some weird bug). Built on Ubuntu 14:04, though not tested extensively yet. I imagine that as I continue my work through the summer, I will expand the image.

Check it out: https://github.com/sirgogo/docker-meep
